### PR TITLE
Add Proxy Address/Group/Policy cmdlets

### PR DESCRIPTION
**Pull request by Tuurtje**

### CODE REVIEW

File: PowerFGT/Private/Confirm.ps1 [hunk 1]
```diff
@@ -63,6 +63,65 @@ Function Confirm-FGTAddressGroup {
...
 }
 
+Function Confirm-FGTProxyAddress {
+
+    Param (
+        [Parameter (Mandatory = $true)]
+        [object]$argument
+    )
+
+    #Check if it looks like an Address element
```
Comment (alagoutte) [inline PowerFGT/Private/Confirm.ps1]: Address => ProxyAddress

File: PowerFGT/Private/Confirm.ps1 [hunk 2]
```diff
@@ -63,6 +63,65 @@ Function Confirm-FGTAddressGroup {
...
+}
+
+Function Confirm-FGTProxyAddressGroup {
+
+    Param (
+        [Parameter (Mandatory = $true)]
+        [object]$argument
+    )
+
+    #Check if it looks like an Address element
```
Comment (alagoutte) [inline PowerFGT/Private/Confirm.ps1]: Proxy Address Group

File: PowerFGT/Private/Confirm.ps1 [hunk 3]
```diff
@@ -104,6 +163,47 @@ Function Confirm-FGTFirewallPolicy {
...
 }
 
+Function Confirm-FGTFirewallProxyPolicy {
+
+    Param (
+        [Parameter (Mandatory = $true)]
+        [object]$argument
+    )
+
+    #Check if it looks like an Firewall Policy element
```
Comment (alagoutte) [inline PowerFGT/Private/Confirm.ps1]: FirewallProxyPolicy ? (it the same like Firewall Policy ?)

File: PowerFGT/Public/cmdb/firewall/proxy-address.ps1 [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+
+    <#
+        .SYNOPSIS
+        Add a FortiGate ProxyAddress
+
+        .DESCRIPTION
+        Add a FortiGate ProxyAddress (host-regex, url, category...)
+
+        .EXAMPLE
+        Add-FGTFirewallAddress -type ipmask -Name FGT -ip 192.2.0.0 -mask 255.255.255.0
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/firewall/proxy-address.ps1]: Wrong Example ?

File: PowerFGT/Public/cmdb/firewall/proxy-address.ps1 [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+
+    #>
+    [CmdletBinding(PositionalBinding=$false)]
+    Param(
+        [Parameter (Mandatory = $true)]
+        [ValidateSet("host-regex", "url", "method","category",IgnoreCase = $false)]
+        [string]$type,
+        [Parameter (Mandatory = $true)]
+        [string]$name,
+
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/firewall/proxy-address.ps1]: remove blank line

File: PowerFGT/Public/cmdb/firewall/proxy-address.ps1 [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+        [ValidateLength(0, 255)]
+        [string]$comment,
+        [Parameter (Mandatory = $false)]
+        [boolean]$visibility,
+        [Parameter(Mandatory = $false)]
+        [String[]]$vdom,
+        [Parameter(Mandatory = $false)]
+        [psobject]$connection = $DefaultFGTConnection
+    )
+DynamicParam
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/firewall/proxy-address.ps1]: really need to use DynamicParam ?

File: PowerFGT/Public/cmdb/firewall/proxy-policy.ps1 [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+
+    <#
+        .SYNOPSIS
+        Add a FortiGate Policy
+
+        .DESCRIPTION
+        Add a FortiGate Policy/Rules (source port/ip, destination port, ip, action, status...)
+
+        .EXAMPLE
+        Add-FGTFirewallPolicy -name MyFGTPolicy -srcintf port1 -dstintf port2 -srcaddr all -dstaddr all
```
Comment (alagoutte) [inline PowerFGT/Public/cmdb/firewall/proxy-policy.ps1]: same here about EXAMPLE

File: PowerFGT/Private/Confirm.ps1 (hunk 1 above)
Comment (Tuurtje) [reply PowerFGT/Private/Confirm.ps1]: updated

File: PowerFGT/Private/Confirm.ps1 (hunk 2 above)
Comment (Tuurtje) [reply PowerFGT/Private/Confirm.ps1]: updated

File: PowerFGT/Private/Confirm.ps1 (hunk 3 above)
Comment (Tuurtje) [reply PowerFGT/Private/Confirm.ps1]: It looks the same, but there is a small difference. There is a check for:
$argument | get-member -name **proxy** -Membertype Properties)

File: PowerFGT/Public/cmdb/firewall/proxy-address.ps1 (hunk 4 above)
Comment (Tuurtje) [reply PowerFGT/Public/cmdb/firewall/proxy-address.ps1]: Updated all examples en tested them. Removed Category for now (pretty complex)

File: PowerFGT/Public/cmdb/firewall/proxy-address.ps1 (hunk 5 above)
Comment (Tuurtje) [reply PowerFGT/Public/cmdb/firewall/proxy-address.ps1]: Done

File: PowerFGT/Public/cmdb/firewall/proxy-address.ps1 (hunk 6 above)
Comment (Tuurtje) [reply PowerFGT/Public/cmdb/firewall/proxy-address.ps1]: As discussed before, its not needed but usefull. It is also used to collect the validationset. The validationset might take up a long time when there are many objects. i only tested it on a test fortigate. Maybe not use it in large production systems.

File: PowerFGT/Public/cmdb/firewall/proxy-policy.ps1 (hunk 7 above)
Comment (Tuurtje) [reply PowerFGT/Public/cmdb/firewall/proxy-policy.ps1]: Updated

**Comment by Tuurtje** [comment]
I removed Set-FGTFirewallProxyAddress function, because it was not ready yet. I will look in to that when i got the time.

File: PowerFGT/Public/cmdb/firewall/proxy-address.ps1 (hunk 6 above)
Comment (alagoutte) [reply PowerFGT/Public/cmdb/firewall/proxy-address.ps1]: i plan to a next release to add better validation (with cache)
can you use ParametSet for the moment ? (like other module

**Comment by alagoutte** [comment]
> I removed Set-FGTFirewallProxyAddress function, because it was not ready yet. I will look in to that when i got the time.

Sure, no problem ! (it is may be like SetFGTFirewallAddress) if the object is use... (you can not modified all parameter...)
it is on my TODO too ;-) (when after i will write tests...)

**Comment by Tuurtje** [comment]
> Hi Arthur,
> 
> Sorry for delay, but i found time to finish test (for Proxy Policy), can you look before i will merge ?

Thanks for all the code changes. I went through all the changes you made and it looks great en very clean. I think its ready to merge.

**Comment by alagoutte** [comment]
Ok found a issue... with FortiGate 6.0.. need to enable Explicit Proxy before add proxy explicit policy 
i will see for fix this issue too

**Pull request merged**